### DISCUSSION TITLE
[Question] How to add frames ?

### DISCUSSION THREAD

**Discussion by eserdeiro** [hm21/pro_image_editor#246] [Q&A] [answered]
I would like to implement a feature that allows adding customizable frames to a photo, similar to the examples in the attached image (with balloons, banners, and celebration themes). How can I achieve this in my project?

Some key functionalities I am looking for:

Ability to choose from different frame designs.
Option for the user to upload their photo after selecting the frame.
Option to export the final image with the applied frame.

Any guidance or examples would be appreciated.
![Captura de pantalla 2024-10-16 a la(s) 1 10 33 p  m](https://github.com/user-attachments/assets/2c9af1cd-0f78-4091-88fa-02732b371b2a)

**Comment by hm21** [comment]
Hi,

I’ll move your questions from the Issues tab to the Discussions tab, as I prefer having questions there. It’s better for Q&A, making it easier for other users to find answers to similar questions.


Anyway, that's an interesting feature! To provide the best solution, I have a few questions:

1. Should the frames have a fixed position, or should users be able to move and resize them?
2. Do users need to perform other actions, like painting or cropping the image?
3. Would it make more sense for users to upload the image first and then select the frame, so they can immediately see how it looks?

**Reply by eserdeiro** [reply]
Thank you very much for your time!

To clarify your questions:

The frames should have a fixed position since they will define the aspect ratio of the final image (as in the example, it is 16x9).
Each frame should allow the user to insert images into the “slots” (in this case, 3 slots).
<img src="https://github.com/user-attachments/assets/cb3d5410-0039-4b82-904d-16745d3bbca3" alt="1" width="300"/>

The images inside these slots should be adjustable, allowing users to rotate or resize them using a two-finger gesture, similar to how emojis are implemented.
<img src="https://github.com/user-attachments/assets/c97824a8-fb6e-42d9-930b-549dd2a64d2a" alt="2" width="300"/>

I’d also like frames to be insertable from a list of .png URLs, similar to how emojis are inserted.
To summarize:

Initially, the user should choose whether they want frames or not.
If they don’t want frames, they proceed to the regular photo editor.
If they choose frames, they pick a frame and upload their photos, then move to the conventional photo editor where they can further edit the content.


Here’s an example of a frame .png:
<img src="https://github.com/user-attachments/assets/70f32027-dd78-45bc-b174-6e134aaf783e" alt="3" width="400"/>

Again, I appreciate your attention to this matter!


Best regards!

**Reply by lcundiff** [reply]
Oh these are like canva frames where the image is sort of under the layer of the frame. We added something similar to ours by making a copy of the sticker feature to make a "template" feature and setting the sticker to the size of the canvas. So you could do this with an image that can be a png with part of it transparent, then when user adds content, they can set it behind for similar effect above. This would only work when the frame takes up the entire canvas though, since if it didn't the image would poke out from behind the frame. 


Also it would be easier if you could just click on each frame and insert images/sticker in them where the photo gets somehow hidden from outside the frame, but that would require changes to the library I imagine. Could maybe work with rectangular frames by making it a clickable sticker that prompts and image and sticks the image below the frame and crops it the size of the frame?

**Reply by hm21** [reply]
@eserdeiro, I apologize for not seeing your response to my questions sooner. As @lcundiff mentioned, this will require some changes to the library. I’ve already implemented these changes in version `6.1.0` and included an [example](https://github.com/hm21/pro_image_editor/blob/stable/example/lib/pages/frame_example.dart) to illustrate how frames can be managed. The only part you’ll need to handle yourself is the logic for where users can insert their images.

If you have a few frames, the simplest approach would be to create a hardcoded solution, with a designated button for each frame to insert an image.

However, if you have a larger number of frames, I recommend writing a function that detects transparent areas and automatically places the buttons at the center of each area. If you plan to do this at runtime, consider using the [image](https://pub.dev/packages/image) package. You could implement a method similar to what I’ve done [here](https://github.com/hm21/pro_image_editor/blob/01c7092b8aa31c4ce22d673c6427579530085094/lib/utils/content_recorder.dart/utils/dart_ui_remove_transparent_image_areas.dart#L20-L103) to identify transparent regions.

Keep in mind, though, that performing this detection on the client side may impact performance, so a server-side implementation may be preferable.

**Reply by eserdeiro** [reply, marked as answer]
Hi! I've made progress on this over the week. 
It's still in development, but there's a functional foundation that should be helpful. 
It meets the initial requirements; I just need to add the image picker.

<img src="https://github.com/user-attachments/assets/cacd9bf7-049b-4e16-b6ce-a33f0e45d217" alt="2" width="300"/>
<img src="https://github.com/user-attachments/assets/de72a08a-2bbc-49b0-aab8-589a3aa82cac" alt="2" width="300"/>

```dart
import 'dart:developer';
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:image/image.dart' as img;

class AreaDetection extends StatefulWidget {
  const AreaDetection({super.key});

  @override
  AreaDetectionState createState() => AreaDetectionState();
}

class AreaDetectionState extends State<AreaDetection> {
  List<Map<String, double>> transparentAreas = [];
  Uint8List? imageData;

  Future<void> _loadAndDetectTransparentAreas() async {
    final response = await http.get(
      Uri.parse(
        'YOUR_FRAME_PATH',
      ),
    );
    if (response.statusCode == 200) {
      imageData = response.bodyBytes;
      transparentAreas = await detectLargeTransparentAreas(imageData!);
    } else {
      log('Error downloading image');
    }
  }

  Map<String, int> floodFill(
    img.Image image,
    int startX,
    int startY,
    List<List<bool>> visited, {
    int horizontalExpansion = 5,
  }) {
    final stack = <Map<String, int>>[];
    var minX = startX;
    var minY = startY;
    var maxX = startX;
    var maxY = startY;
    var size = 0;

    stack.add({'x': startX, 'y': startY});
    while (stack.isNotEmpty) {
      final point = stack.removeLast();
      final x = point['x']!;
      final y = point['y']!;

      if (x < 0 || x >= image.width || y < 0 || y >= image.height) continue;
      if (visited[y][x] || image.getPixel(x, y).a != 0) continue;

      visited[y][x] = true;
      size++;

      minX = x < minX ? x : minX;
      maxX = x > maxX ? x : maxX;
      minY = y < minY ? y : minY;
      maxY = y > maxY ? y : maxY;

      stack
        ..add({'x': x + 1, 'y': y})
        ..add({'x': x - 1, 'y': y})
        ..add({'x': x, 'y': y + 1})
        ..add({'x': x, 'y': y - 1});
    }

    minX = (minX - horizontalExpansion).clamp(0, image.width - 1);
    maxX = (maxX + horizontalExpansion).clamp(0, image.width - 1);

    return {
      'minX': minX,
      'minY': minY,
      'maxX': maxX,
      'maxY': maxY,
      'size': size,
    };
  }

  Future<List<Map<String, double>>> detectLargeTransparentAreas(
    Uint8List imageData, {
    int minAreaSize = 1000,
  }) async {
    final image = img.decodeImage(imageData);
    final areas = <Map<String, double>>[];
    if (image == null) return areas;

    final visited = List.generate(
      image.height,
      (_) => List<bool>.filled(image.width, false),
    );

    for (var y = 0; y < image.height; y++) {
      for (var x = 0; x < image.width; x++) {
        final pixel = image.getPixel(x, y);
        if (pixel.a == 0 && !visited[y][x]) {
          final area = floodFill(image, x, y, visited);
          if (area['size']! >= minAreaSize) {
            final centerX = (area['minX']! + area['maxX']!) / 2;
            final centerY = (area['minY']! + area['maxY']!) / 2;

            areas.add({
              'centerX': centerX / image.width,
              'centerY': centerY / image.height,
              'width': (area['maxX']! - area['minX']!) / image.width,
              'height': (area['maxY']! - area['minY']!) / image.height,
            });
          }
        }
      }
    }
    return areas;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Frames')),
      body: FutureBuilder<void>(
        future: _loadAndDetectTransparentAreas(),
        builder: (context, snapshot) {
          if (snapshot.connectionState == ConnectionState.waiting) {
            return const Center(child: CircularProgressIndicator());
          } else if (snapshot.hasError) {
            return const Center(child: Text('Error downloading image'));
          } else {
            return LayoutBuilder(
              builder: (context, constraints) {
                final imageWidth = constraints.maxWidth;
                final imageHeight = constraints.maxHeight * 0.7;
                final originalImage = img.decodeImage(imageData!);

                if (originalImage == null) {
                  return const Text('Image error');
                }

                final imageAspectRatio =
                    originalImage.width / originalImage.height;
                final containerAspectRatio = imageWidth / imageHeight;

                var offsetX = 0;
                var offsetY = 0;
                double scale;

                if (containerAspectRatio > imageAspectRatio) {
                  scale = imageHeight / originalImage.height;
                  offsetX =
                      ((imageWidth - originalImage.width * scale) / 2).toInt();
                } else {
                  scale = imageWidth / originalImage.width;
                  offsetY = ((imageHeight - originalImage.height * scale) / 2)
                      .toInt();
                }

                return Column(
                  children: [
                    Stack(
                      children: [
                        for (final area in transparentAreas)
                          Positioned(
                            left: offsetX +
                                (area['centerX']! *
                                    originalImage.width *
                                    scale) -
                                (area['width']! * originalImage.width * scale) /
                                    2,
                            top: offsetY +
                                (area['centerY']! *
                                    originalImage.height *
                                    scale) -
                                (area['height']! *
                                        originalImage.height *
                                        scale) /
                                    2,
                            child: Container(
                              width:
                                  area['width']! * originalImage.width * scale,
                              height: area['height']! *
                                  originalImage.height *
                                  scale,
                              color: Colors.red.withOpacity(.9),
                            ),
                          ),
                        Image.memory(
                          imageData!,
                          width: imageWidth,
                          height: imageHeight,
                          fit: BoxFit.contain,
                        ),
                        for (final area in transparentAreas)
                          Positioned(
                            left: offsetX +
                                (area['centerX']! *
                                    originalImage.width *
                                    scale) -
                                30,
                            top: offsetY +
                                (area['centerY']! *
                                    originalImage.height *
                                    scale) -
                                24,
                            child: IconButton(
                              onPressed: () {
                                log('On pressed');
                              },
                              icon: const Icon(Icons.add_circle),
                              iconSize: 48,
                            ),
                          ),
                      ],
                    ),
                  ],
                );
              },
            );
          }
        },
      ),
    );
  }
}
```

**Reply by hm21** [reply]
@eserdeiro Your implementation looks good for detecting transparent areas—great job! I have just a small suggestion: in cases where users may be on lower-performance devices or when processing high-resolution images, such as 4K, you might want to consider using a web worker on dart-web or on dart-native isolating the processing. This approach can help prevent the UI from freezing during this intensive task.
